mobile-browser-based-version/src/helpers/training/training_manager.js

```diff
@@ -144,79 +150,263 @@ export class TrainingManager {
+    var dataSize = -1;
+
+    // If info resize is not defined it will default to Fast Training
+    if (info.resize) {
+      // 8 is for 8 bytes in a single element and 3 is for the 3 color channels for image tasks
+      dataSize =
+        8 *
+        3 *
+        info.RESIZED_IMAGE_H *
+        info.RESIZED_IMAGE_W *
+        this.data.shape[0];
+    }
+
+    if (dataSize > SIZE_THRESHOLD) {
+      console.log('Memory Efficient training');
+      // Creation of Dataset objects for training
+      const trainData = tf.data
+        .generator(
+          imageDatasetGenerator(
+            this.data,
+            this.labels,
+            0,
+            this.data.shape[0] * (1 - info.validationSplit),
+            (tensor) =>
+              tf.image.resizeBilinear(tensor, [
+                info.RESIZED_IMAGE_H,
+                info.RESIZED_IMAGE_W,
+              ])
+          )
+        )
+        .batch(info.batchSize);
+
+      const valData = tf.data
+        .generator(
+          imageDatasetGenerator(
+            this.data,
+            this.labels,
+            Math.floor(this.data.shape[0] * (1 - info.validationSplit)),
+            this.data.shape[0],
+            (tensor) =>
+              tf.image.resizeBilinear(tensor, [
+                info.RESIZED_IMAGE_H,
+                info.RESIZED_IMAGE_W,
+              ])
+          )
+        )
+        .batch(info.batchSize);
+
+      await this.model.fitDataset(trainData, {
+        epochs: info.epochs,
+        validationData: valData,
+        callbacks: {
+          onEpochEnd: async (epoch, logs) => {
+            this.trainingInformant.updateGraph(
+              epoch + 1,
+              (logs.val_acc * 100).toFixed(2),
+              (logs.acc * 100).toFixed(2)
+            );
+            console.log(
+              `EPOCH (${epoch + 1}):
             Train Accuracy: ${(logs.acc * 100).toFixed(2)},
             Val Accuracy:  ${(logs.val_acc * 100).toFixed(2)}\n`
-          );
-          if (this.useIndexedDB) {
-            this.model.setUserDefinedMetadata({ epoch: epoch + 1 });
-            await updateWorkingModel(
-              this.task.taskID,
-              info.modelID,
-              this.model
             );
-          }
-          if (this.stopTrainingRequested) {
-            this.model.stopTraining = true;
-            this.stopTrainingRequested = false;
-          }
+            if (this.useIndexedDB) {
+              this.model.setUserDefinedMetadata({ epoch: epoch + 1 });
+              await updateWorkingModel(
+                this.task.taskID,
+                info.modelID,
+                this.model
+              );
+            }
+            if (this.stopTrainingRequested) {
+              this.model.stopTraining = true;
+              this.stopTrainingRequested = false;
+            }
+          },
         },
-      },
-    });
-  }
+      });
+    } else {
+      console.log('Fast training');
+
+      let resized_data = this.data;
```

Comment:
resizing is a bit too specific, resizing can be just part of some general preprocessing anyway (no matter if batch-wise or done globally before training)?